### ISSUE TITLE
On unifying Array into AbstractArray

### ISSUE THREAD

**Issue by vavrines** [timueh/PolyChaos.jl#26]
Under the current function declarations of Array, it may bring some inconvenience when we try to include more usage scenarios with arrays affiliated to AbstractArray. See an example.

First let's define a variable in polynomial chaos expansion with type of OffsetArray.
```
using PolyChaos
L = 5; op = GaussOrthoPoly(L, addQuadrature=true)

using OffsetArrays
nx = 10; vChaos = OffsetArray{Float64}(undef, 0:nx+1, 1:L+1); vChaos .= 1.
```

We would like to evaluate the PCE onto quadrature points.
```
vQuad = OffsetArray{Float64}(undef, 0:nx+1, 1:op.quad.Nquad)
for i in axes(pce)[1]
    vQuad[i,:] = evaluatePCE(vChaos[i,:], op.quad.nodes, op)
end
```
Unfortunately, here it should turn out:
```
MethodError: no method matching evaluatePCE(::OffsetArray{Float64,1,Array{Float64,1}}, ::Array{Float64,1}, ::GaussOrthoPoly)
Closest candidates are:
  evaluatePCE(!Matched::Array{#s99,1} where #s99<:Real, ::Array{#s98,1} where #s98<:Real, ::AbstractOrthoPoly) at /home/tianbai/.julia/packages/PolyChaos/1jqDA/src/polynomial_chaos.jl:229
  evaluatePCE(!Matched::Array{#s99,1} where #s99<:Real, ::Array{#s98,1} where #s98<:Real, !Matched::Array{#s97,1} where #s97<:Real, !Matched::Array{#s96,1} where #s96<:Real) at /home/tianbai/.julia/packages/PolyChaos/1jqDA/src/polynomial_chaos.jl:216

Stacktrace:
 [1] top-level scope at ./In[17]:9

```
It is due to the incompatibility in the function declarations. A temporary solution is to reformulate an array from AbstractArray, then put it into functions.
```
vQuad1 = OffsetArray{Float64}(undef, 0:nx+1, 1:op.quad.Nquad)
for i in axes(pce)[1]
    vQuad1[i,1:op.quad.Nquad] = evaluatePCE(vChaos[i,1:L+1], op.quad.nodes, op)
end

vQuad1
12×5 OffsetArray(::Array{Float64,2}, 0:11, 1:5) with eltype Float64 with indices 0:11×1:5:
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
 11.2059  -2.5914  3.0  -3.03138  46.4168
```

A more general solution is to modify the function declarations correspondingly.

**Comment by timueh** [comment]
Hey @vavrines, I am working on it as you can tell. I am hoping to get it done before Christmas.

Weirdly enough, the build fails.. but that's an issue I'll definitely need to consult discourse.julialang for.. unless you have ideas?!..

**Comment by timueh** [comment]
This is working now for me. But.. it's still WIP!

> using PolyChaos, OffsetArrays
> L = 5; op = GaussOrthoPoly(L, addQuadrature=true)
> 
> using OffsetArrays
> nx = 10; vChaos = OffsetArray{Float64}(undef, 0:nx+1, 1:L+1); vChaos .= 1.



> vQuad = OffsetArray{Float64}(undef, 0:nx+1, 1:op.quad.Nquad)
> for i in axes(__vChaos__)[1]
>     vQuad[i,:] = evaluatePCE(vChaos[i,:], op.quad.nodes, op)
> end

**Comment by timueh** [comment]
Hey @vavrines, I fixed the build issues. Feel free to test it. Feedback is much appreciated.

**Comment by timueh** [comment]
Hey @vavrines, have you had time to play with it?

**Comment by vavrines** [comment]
> Hey @vavrines, have you had time to play with it?

Thanks and sorry for a late reply. I'm just back from vacation.
Maybe a noob question, but I'm wondering which branch should I go to test this?
I tried `add PolyChaos#master` and test it, but it turns to show the same error.

**Comment by timueh** [comment]
The branch is [`use_abstract_arrays`](https://github.com/timueh/PolyChaos.jl/tree/use_abstract_arrays).

I should've mentioned that. :smile:

**Comment by vavrines** [comment]
> The branch is [`use_abstract_arrays`](https://github.com/timueh/PolyChaos.jl/tree/use_abstract_arrays).
> 
> I should've mentioned that. 😄

Thanks! I did some preliminary tests and this branch worked well.

**Comment by timueh** [comment]
Good to know.

I learned yesterday [that entries should have concrete types](https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations-1). That means, I need to modify the type definitions, e.g.

https://github.com/timueh/PolyChaos.jl/blob/00a4f1b36cbfd7b7aca59d21f13130ca6df71560/src/typesMeasures.jl#L19-L29

So if you want to use it, just use the branch. I can't say at the moment when I'll get around changing the type definitions.

**Comment by vavrines** [comment]
> Good to know.
> 
> I learned yesterday [that entries should have concrete types](https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations-1). That means, I need to modify the type definitions, e.g.
> 
> https://github.com/timueh/PolyChaos.jl/blob/00a4f1b36cbfd7b7aca59d21f13130ca6df71560/src/typesMeasures.jl#L19-L29
> 
> So if you want to use it, just use the branch. I can't say at the moment when I'll get around changing the type definitions.

Thanks for this interesting knowledge.
But I'm curious that how much performance will be enhanced after this treatment.
In real computations, the Measure could only need to be defined once, but it's more time consuming for matrix operations and numerical quadrature.

**Comment by timueh** [comment]
I believe I fixed it with #b490512. Feel free to take a look. I guess, I'll release it as a new version soon.

**Comment by timueh** [comment]
The [new version](https://github.com/timueh/PolyChaos.jl/releases) of `PolyChaos` is out.
